### PR TITLE
IDENTITY-5185:OIDC DCR does not validate the request and handle the errors

### PR DESCRIPTION
**Pull request by Abilashini**
Validate the redirect URI in the request.
Response with an error if the redirect URI is not valid.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java [hunk 1]
```diff
@@ -117,6 +119,13 @@ public void create(IdentityRequest.IdentityRequestBuilder builder, HttpServletRe
...
                 JSONArray redirectUris = (JSONArray) obj;
                 for (int i = 0; i < redirectUris.size(); i++) {
+                    try{
+                        URL redirectURL = new URL(redirectUris.get(i).toString());
+                    }
+                    catch (MalformedURLException e){
+                        String errorMessage = "The redirection URI " + redirectUris.get(i).toString() + " is not allowed by this Server";
+                        throw IdentityException.error(FrameworkClientException.class, errorMessage, e);
+                    }
                     registrationRequestProfile.getRedirectUris().add(redirectUris.get(i).toString());
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java]: Can improve the code by defining a variable for 'redirectUris.get(i).toString()'  and using same in both the places line 123 and 129.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java [hunk 2]
```diff
@@ -117,6 +119,13 @@ public void create(IdentityRequest.IdentityRequestBuilder builder, HttpServletRe
             if (obj != null && obj instanceof JSONArray) {
                 JSONArray redirectUris = (JSONArray) obj;
                 for (int i = 0; i < redirectUris.size(); i++) {
+                    try{
+                        URL redirectURL = new URL(redirectUris.get(i).toString());
+                    }
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java]: @Pushpalanka, We need to stick to correct code formatting.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java [hunk 3]
```diff
@@ -117,6 +119,13 @@ public void create(IdentityRequest.IdentityRequestBuilder builder, HttpServletRe
...
+                    catch (MalformedURLException e){
+                        String errorMessage = "The redirection URI " + redirectUris.get(i).toString() + " is not allowed by this Server";
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java]: @Pushpalanka The error message is misleading, better to say 
"The redirection URI --- could not be converted to URL"

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java [hunk 4]
```diff
@@ -117,6 +119,13 @@ public void create(IdentityRequest.IdentityRequestBuilder builder, HttpServletRe
             if (obj != null && obj instanceof JSONArray) {
                 JSONArray redirectUris = (JSONArray) obj;
                 for (int i = 0; i < redirectUris.size(); i++) {
+                    try{
+                        URL redirectURL = new URL(redirectUris.get(i).toString());
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java]: @Pushpalanka Need to test if the Array element is null (i.e. JSON null) before doing anything else. Some points of doing so is,
http://stackoverflow.com/questions/12251729/how-to-check-if-jsonarray-element-is-null

Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/factory/RegistrationRequestFactory.java]: According to the relevant specifications, these are URIs. Hence it is wrong to validate them to be URLs. We have correct that to valid to be URIs.

**Comment by Pushpalanka** [comment]
Review comments addressed at #333
